### ISSUE TITLE
Force reload on app upgrade - possible new header

### ISSUE THREAD

**Issue by pfw** [unpoly/unpoly#126]
I've run across what appears to be a quite elegant solution to a problem that can sometimes be hit as new versions of applications are hit where a user is mid session as the upgrade is pushed out and the new pages have conflicts in the DOM.

The idea would be to allow for sending a 'version' header back from the server on each request and then having Unpoly send that on each subsequent request. The server can then check to see if the app version has changed and would signal to Unpoly to perform a full reload of a given URL.

"Asset versioning" on https://inertiajs.com/the-protocol describes the process that looks nice - it is fairly straightforward to implement both on the client and server. I suspect this could be made to work at the moment through watching up:proxy:load/up:proxy:loaded, setting headers and setting window.location if required but would be nice to be a builtin as well.

**Comment by pfw** [comment]
The following bits do the job but would be nice to not require compilers:

Set on initial load:
``` html
 <meta name="X-Up-AssetsVersion" content="47692">
```

Compilers to inspect load/loaded events:
``` javascript
up.on(document, 'up:proxy:load', (event, element) => {
    let metaVersion = document.querySelector("meta[name=X-Up-AssetsVersion]");
    if (metaVersion) {
        console.log(`Set version header == ${metaVersion.content}`);
        event.request.headers['X-Up-AssetsVersion'] = metaVersion.content;
    }
});

up.on(document, 'up:proxy:loaded', (event, element) => {
    if (event.response.status === 409) {
        let upLocation = event.response.getHeader('X-Up-Location');
        if (!upLocation) {
            console.log("Version conflict signalled but no location specified");
        } else {
            console.log(`Version conflict signalled by server, reload at ${upLocation}`)
            event.preventDefault();
            window.location = upLocation;
        }
    }
});
```

And a python server check of:

``` python
    if request.headers.get("X-Up-AssetsVersion", version) != version:
        print(f"Version mismatch, send 409, cause unpoly to reload")
        return make_response("Asset Versions Conflict", 409)
```

I'm attempting to learn some Coffeescript so I can have a look at doing some kind of pull request if the idea is appealing.

**Comment by triskweline** [comment]
I would like to cover this case in Unpoly, since it affects every web application.

See [related discussion](https://groups.google.com/forum/#!topic/unpoly/UhJp-6j0ySw) about the upcoming `up:fragment:loaded` event that will make this much easier to implement. This new event will be included in the next version of Unpoly.

> The following bits do the job

Thanks for posting working code for this!

Is there a way we can make this new protocol easier to implement by server-side integration libraries like `unpoly-rails` or `ex_unpoly`? Ideally these libraries should require no additional configuration by the user. Since we're reading the initial asset version from a `<meta>` tag, the user will need to manually change her application layout the include that new tag. It's really hard for a library like unpoly-rails to automatically insert this tag into the response body.

What's really simple for a library like unpoly-rails is to manipulate response headers. So I'm proposing this protocol:

- When the server sees an Unpoly request, it adds a response header `X-Up-Asset-Version: 1234`.
- When the server sees a non-Unpoly request, it sets a temporary cookie `X-Up-Asset-Version: 1234`. This is a workaround for there no being no JavaScript API to read the response header on the initial page load.

We could then use the new `up:fragment:loaded` event to implement the feature like this:

```javascript
// The last seen asset version, initially set to the version from the original page load.
// Because there is no JavaScript API to read the response header of the original page load,
// the server sets a cookie that we immediately read and delete.
let previousAssetVersion = up.browser.popCookie('X-Up-Asset-Version')

up.on('up:fragment:loaded', (event) => {
  let fragmentAssetVersion = event.response.getHeader('X-Up-Asset-Version')
  
  // Do nothing if the server didn't send an asset version
  if (!fragmentAssetVersion) return
  
  // Check if the asset version has changed
  if (previousAssetVersion && previousAssetVersion != fragmentAssetVersion) {
    // Prevent the fragment update and don't update browser history
    event.preventDefault()    

    // Make a full page load for the same request.
    event.request.navigate()
  }

  // Remember the response's asset version for the next response.
  // In case the server didn't set the X-Up-Asset-Version cookie in the initial page
  // load we know have a version to compare.
  previousAssetVersion = fragmentAssetVersion;
})
```

> I'm attempting to learn some Coffeescript so I can have a look at doing some kind of pull request if the idea is appealing.

A PR for this would be great.

I given that the next release includes `up:fragment:loaded` and changes many internals, I recommend that you either wait for the next Unpoly 1.0.0 release or target the hk/layers branch in which the release is being prepared.

**Comment by pfw** [comment]
That all makes sense to me. I'll see if I can get a working dev environment for the hk/layers branch.

**Comment by triskweline** [comment]
I realized that tests were broken in `hk/layers`. This is fixed now.

Let me know if you need help writing tests. Writing a Jasmine spec that involves mocked AJAX requests may be daunting when you do it the first time.

**Comment by pfw** [comment]
Thanks. I've got what looks like a working ruby/rails dev environment... I think, I'll give it a go.

**Comment by johndwells** [comment]
If Unpoly is going to provide a way to handle this in core, I don't think it should be relying on a cookie at least not without an alternative way to set the current version. There are cases where a server response may not send any cookies, such as a statically cached response.

Either setting the initial version via JS (e.g. `up.proxy.config.assetVersion = '1234';`) or via a meta tag would be ideal IMHO.

FWIW, our own version of this logic looks like:

```
// config.version has been set by our own code earlier up
up.on('up:proxy:loaded', function(event) {
  let responseVersion = event.response.getHeader('X-Up-Version');
  if(responseVersion !== config.version) {
    up.log.debug('New version detected: ' + responseVersion);
    event.preventDefault();
    up.proxy.clear();
    up.history.config.enabled = false;
    return event.request.navigate();
  }
});
```

**Comment by triskweline** [comment]
> Either setting the initial version via JS (e.g. up.proxy.config.assetVersion = '1234';) or via a meta tag would be ideal IMHO.

I like `up.proxy.config.assetVersion` as an alternative to the cookie! I still like to allow the cookie for its potential to be zero-conf.

> FWIW, our own version of this logic looks like:

Two comments regarding the code snippet:

- We should not rely on the initial asset version to be defined (we cannot force other developers to set the cookie or up.proxy.config.assetVersion)
- As per [yesterday](https://github.com/unpoly/unpoly/commit/1103e53b9caf011453b59b6167dc1f3b616bdbf4) `request.navigate()` will abort all pending requests, so we shouldn't need to disable history or fear other requests from resolving while we're navigating away.

**Comment by johndwells** [comment]
@triskweline Thanks for the tip on `request.navigate()`!

**Comment by triskweline** [comment]
> Thanks for the tip on request.navigate()!

It's not released yet, so don't change your code yet :)

**Comment by johndwells** [comment]
Right! Got excited there. :)

**Comment by pfw** [comment]
Life got in the way of work on this for me but I should be able to get back shortly - I just noticed commits that show 1.0 being close, I'll take a pass at this again once it is all set.

**Comment by triskweline** [comment]
No worries, thanks for the update.

**Comment by triskweline** [comment]
The use case described in this use is solved by Unpoly 3.5.0.
See [Handling changes in frontend code](https://unpoly.com/handling-asset-changes).
